### PR TITLE
Async envent processing

### PR DESCRIPTION
**Pull request by janisz**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -132,9 +133,10 @@ metrics-location       |                       | Graphite URL (used when metrics
 metrics-prefix         | `default`             | Metrics prefix (resolved to `<hostname>.<app_name>` by default)
 metrics-target         | `stdout`              | Metrics destination `stdout` or `graphite` (empty string disables metrics)
 sync-enabled           | `true`                | Enable Marathon-consul scheduled sync
+sync-force             | `false`               | Force leadership-independent Marathon-consul sync (run always)
 sync-interval          | `15m0s`               | Marathon-consul sync [interval](https://golang.org/pkg/time/#Duration) **Note:** While using file configuration intervals should be provided in *nanoseconds*
 sync-leader            | `<hostname>:8080`     | Marathon cluster-wide node name (defaults to `<hostname>:8080`), the sync will run only if the node is the current [Marathon-leader](https://mesosphere.github.io/marathon/docs/rest-api.html#get-v2-leader)
-sync-force             | `false`               | Force leadership-independent Marathon-consul sync (run always)
+workers-pool-size      |:`10`                  | Number of concurrent workers processing events
```
Comment (dankraw) [inline README.md]: :10 -> 10

File: README.md [hunk 2]
```diff
@@ -119,6 +119,7 @@ consul-token           |                       | The Consul ACL token
 consul-tag             | `marathon`            | Common tag name added to every service registered in Consul, should be unique for every Marathon-cluster connected to Consul
 consul-timeout         | `3s`                  | Time limit for requests made by the Consul HTTP client. A Timeout of zero means no timeout
 listen                 | `:4000`               | Accept connections at this address
+events-queue-size      | `1000`                | Size of events queue
```
Comment (dankraw) [inline README.md]: maybe move this to bottom?

File: web/event_handler.go [hunk 3]
```diff
@@ -14,68 +14,83 @@ import (
...
+	go func() {
+		for {
+			select {
+			case event = <-fh.eventQueue:
+				metrics.Mark(fmt.Sprintf("events.handler.%d", fh.id))
+				metrics.UpdateGauge("events.queue.len", int64(len(fh.eventQueue)))
+				metrics.UpdateGauge("events.queue.delay_ns", time.Since(event.timestamp).Nanoseconds())
+				metrics.Time("events.processing."+event.eventType, process)
+			case <-quitChan:
+				log.WithField("Id", fh.id).Info("Stoping worker")
```
Comment (dankraw) [inline web/event_handler.go]: Stoping -> Stopping

File: web/event_handler.go [hunk 4]
```diff
@@ -14,68 +14,83 @@ import (
...
+	quitChan := make(chan stopEvent)
+	log.WithField("Id", fh.id).Println("Starting worker")
+	go func() {
+		for {
+			select {
+			case event = <-fh.eventQueue:
+				metrics.Mark(fmt.Sprintf("events.handler.%d", fh.id))
+				metrics.UpdateGauge("events.queue.len", int64(len(fh.eventQueue)))
+				metrics.UpdateGauge("events.queue.delay_ns", time.Since(event.timestamp).Nanoseconds())
+				metrics.Time("events.processing."+event.eventType, process)
```
Comment (dankraw) [inline web/event_handler.go]: nice! 💥

File: web/event_handler.go [hunk 5]
```diff
@@ -85,145 +100,127 @@ func (fh *EventHandler) handleHealthStatusEvent(w http.ResponseWriter, body []by
...
 	}
-	fh.okResponse(w)
 }
 
-/*
-	This handler is used when an application is stopped
-*/
-func (fh *EventHandler) handleDeploymentInfo(w http.ResponseWriter, body []byte) {
-	body = replaceTaskIdWithId(body)
+func mark(task *apps.Task) {
```
Comment (dankraw) [inline web/event_handler.go]: mark -> markStatusEvent

File: web/event_handler.go [hunk 6]
```diff
@@ -85,145 +100,127 @@ func (fh *EventHandler) handleHealthStatusEvent(w http.ResponseWriter, body []by
...
+	log.WithFields(log.Fields{
+		"Id":         task.ID,
+		"TaskStatus": task.TaskStatus,
+	}).Info("Got StatusEvent")
+	app := strings.Trim(strings.Replace(task.AppID.String(), "/", ".", -1), ".")
+	metrics.Mark(fmt.Sprintf("apps.%s.%s", app, task.TaskStatus))
```
Comment (dankraw) [inline web/event_handler.go]: I'm not sure if we need this. There are logs already.

File: web/web.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+	for i := 0; i < config.WorkersCount; i++ {
+		handler := newEventHandler(i, service, marathon, eventQueue)
+		stopChannels[i] = handler.Start()
+	}
+	return newWebHandler(eventQueue).Handle, stop(stopChannels)
+}
+
+func stop(channels []chan<- stopEvent) Stop {
+	return func() {
+		for _, chanel := range channels {
```
Comment (dankraw) [inline web/web.go]: channel or Coco Chanel ;)

File: web/web_handler.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+		metrics.Mark("events.requests." + eventType)
+
+		h.eventQueue <- event{eventType: eventType, body: body, timestamp: time.Now()}
+		accepted(w)
+	})
+}
+
+func handleBadRequest(err error, w http.ResponseWriter) {
+	metrics.Mark("events.response.error.400")
+	w.WriteHeader(http.StatusBadRequest)
```
Comment (dankraw) [inline web/web_handler.go]: I think we should better return only 2xx codes.

Comment (janisz) [inline web/web_handler.go]: 400 is returned only for invalid JSON or when JSON has no `eventType` field

File: web/event_handler_test.go [hunk 9]
```diff
@@ -111,91 +114,100 @@ func TestForwardHandler_NotHandleInvalidEventType(t *testing.T) {
...
 	// then
-	assert.Equal(t, 200, recorder.Code)
-	assert.Len(t, service.RegisteredServicesIds(), 0)
+	assertAccepted(t, recorder)
+	assert.Empty(t, service.RegisteredServicesIds())
+	assert.False(t, marathon.Interactions())
 }
 
-func TestForwardHandler_HandleDeploymentInfoWithStopApplicationActionForMultipleApps(t *testing.T) {
+func TestWebHandler_andleDeploymentInfoWithStopApplicationForOneApp(t *testing.T) {
```
Comment (dankraw) [inline web/event_handler_test.go]: handle

File: web/web_handler.go (hunk 8 above)
Comment (dankraw) [inline web/web_handler.go]: ok thx

File: README.md (hunk 2 above)
Comment (janisz) [inline README.md]: I think it should be above listen. This table is sorted.

Comment (dankraw) [inline README.md]: it's not sorted that much imho :) I think it's related to `workers-pool-size`. Or we should rename `workers-pool-size` to `events-queue-workers-pool-size` to fit better.

File: web/event_handler.go (hunk 6 above)
Comment (janisz) [inline web/event_handler.go]: I'll remove this

File: web/event_handler.go (hunk 5 above)
Comment (janisz) [inline web/event_handler.go]: without logic this function is uneccesary, I'll remove it

Comment (dankraw) [inline web/event_handler.go]: I agree

**Comment by dankraw** [comment]
👍

**Pull request merged**